proposals/ytt/003-openapi-schemas/README.md

```diff
@@ -0,0 +1,80 @@
+
+# Proposal
+Implement a flag to export an [OpenAPI Schema](https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.1.0.md) yaml document generated from a [Data Values schema](https://carvel.dev/ytt/docs/latest/lang-ref-ytt-schema/) file, and an optional Data Values file. When invoking ytt with configuration inputs and the flag `--schema-inspect-openapi`, the result is the OpenAPI formatted document only. 
+
+This flag will behave similar to the `--data-values-inspect` flag, which excludes any templates in the output.
+
+## Examples:
+### Scenario 1: Data Values Schema file only
+`ytt -f schema.yml --schema-inspect-openapi` will generate an OpenAPI Schema document using the values from the schema.yml file.
+
+### Scenario 2: Data Values Schema and Data Values
+`ytt -f schema.yml -f values.yml --schema-inspect-openapi` will generate an OpenAPI Schema document using the merged values of both the schema.yml and the values.yml file.
+
+Note: Using the merged result of both files allows the generated OpenAPI Schema to have a default value for array items, since Data Values Schema files do not yet support default values for arrays. Additionally, it allows the OpenAPI Schema to document the merged result of Data Values Schema and Data Values that will ultimately be used as configuration input during templating.
+
+### Scenario 3: Data Values file only
+`ytt -f values.yml --schema-inspect-openapi` will result in a warning suggesting that a Data Values Schema file be provided. 
+
+The reason for waiting to support Data Values files only is so that the OpenAPI document can have the additional information of default values and descriptions only available in Data Values Schema files. If this functionality comes with no additional work, we should reevaluate this decision.
+
+# Specification
+## Glossary
+* OpenAPI Schema: A document that follows this [spec](https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.1.0.md#openapi-object)
+* Collection: a yaml map or array
+* Configuration inputs: a ytt Schema file and a Data Values file
+* [Data Value Schema](https://carvel.dev/ytt/docs/latest/lang-ref-ytt-schema/) A ytt document that declares default values and type information for parameterized values.
+* [Data Value](https://carvel.dev/ytt/docs/latest/ytt-data-values/) A ytt document that overrides defaults declared in a Data Values Schema.
+
+
+## Generated Openapi Document Spec:
+
+### Root Document fields:
+```yaml
+openapi: 3.1.0 # Version of the openapi spec
+info:
+  version: 1.0.0 # Version of this document
+  title: Openapi schema generated from ytt schema # Title of this document
+components: # Required field that holds 'schemas object'
+  ...
+```
+All openapi schemas generated will be prefixed by the above.
+
+### Openapi Map Example:
+#### Ytt schema:
+```yaml
+#@data/values-schema
+---
+load_balancer:
+  enabled: true
+  static_ip: ""
+```
+#### openapi schema:
+```yaml
+...
+components:
+  schemas:
```

Comment:
since packaging is not expecting such header (and the root document fields) under spec.valuesSchema.openAPIv3 key (similar to CRDs e.g. https://github.com/vmware-tanzu/carvel-kapp-controller/blob/develop/config/crds.yml#L16), we should figure out how to disable/enable that within ytt.

Reply:
I had an assumption that users outside of kapp controller may want the metadata in the header for info like the OpenAPI `version` and `title`, and for using OpenAPI validation tools that require a absolute compliant OpenAPI document, but users may really only need the `components.schemas` section.

A way could be to have a `--schema-inspect-openapi-schema` flag, which only outputs the `components.schemas` section like kapp-controller wants. And then if users need a complete OpenAPI schema with the header metadata, we could use`--schema-inspect-openapi` flag that may be added later. What do you think?

Reply:
I updated the proposal to purpose two flags instead `--schema-inspect-openapi` which would only output the `components.schemas` object, and `--schema-inspect-openapi-metadata`, which would output the entire OpenAPI Doc. We can wait to implement the second flag to hear from users if they need it.

Reply:
It could be useful to consider other potential consumers of this functionality. I'm wondering if this capability should live in ytt vs downstream (e.g. kapp-controller).

Reply:
Consumers of this functionality would use this to document what Data Values inputs a piece of configuration allows, and potentially validate the Data Values against the OpenAPI Spec using a separate tool. Consumers that want to document inputs will put the openapi schema in a place for config consumers to read, for kapp-controller thats in a Kubernetes resource PackageCR, for others it could be on a website, or included with related configuration in a Git repo. 

In those latter cases, having the additional headers makes more sense than the kapp-controller case, since a PackageCR already has the OpenAPI version included, but if it were just a file in a repo that version info would be missing.

Reply:
An update to this proposal includes header information in the `ytt -f config/ --data-values-schema-inspect` command. This follows the conventions other tools use (for example kubectl get --raw openapi/v2) and also makes the Document fully compliant with the OpenAPI Spec. There is another flag `--data-values-schema-inspect -o openapi-schema` which would output only the `components.schemas` section.